lc-compile: Fix possible use of memory before initialisation

peter-b opened the pull request:
Fix several places in `lc-compile` where GCC 6.3.1 detects possible use of uninitialised values in release builds.

livecodeali commented:
@livecode-vulcan review ok e8c9c0c

livecode-vulcan commented:
💙 review by @livecodeali ok e8c9c0c23bc1d8283d50f14a750bbed94ec4d628

livecode-vulcan commented:
💥 merge conflict e8c9c0c23bc1d8283d50f14a750bbed94ec4d628 → 7d11a790b7cced30dde6099d337245a7767acb24

peter-b committed:
lc-compile: Ensure local variable is initialised at acquisition

GCC 6.3.1 reports that the `t_index` variable in
`CountSyntaxNodeMarks()` may be uninitialised in some codepaths.
Since it's initialised in every branch of the switch statement, and
the switch covers every possible `SyntaxNodeKind`, it shouldn't
normally be possible, but it's probably best that the function doesn't
return random numbers on encountering a dodgy syntax node.

peter-b committed:
lc-compile: Prevent unitialized dereferences due to unchecked alloc

In `EmitPushRepeatLabels()` and `EmitAttachRegisterToExpression()`,
the result of `MCMemoryNew()` is not checked.  Ensure that if memory
allocation fails, these functions return `nullptr` rather than an
uninitialized pointer.

peter-b committed:
lc-compile: Ensure file pointer is initialised at declaration

In `GetFilenameOfPosition()`, it is possible for `GetFileOfPosition()`
not to provide a valid `MCFileRef`.  In that case, `GetFileOfPath()`
would dereference uninitialised value as a pointer.

peter-b committed:
lc-compile: Don't return uninitialised pointer from to_mcstringref

If `MCValueInter()` failed, `to_mcstringref()` could return an
uninitialised pointer.  This patch prevents this by using an
`MCAutoStringRef` for the uniqued string.

livecodeali commented:
@livecode-vulcan review ok 26f2817

livecode-vulcan commented:
💙 review by @livecodeali ok 26f2817943804660afc38775e59337934581d6c4

livecode-vulcan commented:
😎 test success 26f2817943804660afc38775e59337934581d6c4

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1019)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1025)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1017)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1018)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1025)
* try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1039)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1016)
